src/importlinter/contracts/layers.py

```diff
@@ -173,6 +187,28 @@ def _check_all_layers_exist_for_container(self, container: str, graph: ImportGra
                     f"module {layer_module_name} does not exist."
                 )
 
+    def _check_all_potential_layers_exist_in_contract(
+        self, container: str, graph: ImportGraph
+    ) -> None:
+        declared_layers = set(layer.name for layer in self.layers)  # type: ignore
+        expected_layers = declared_layers | self.exhaustive_ignores  # type: ignore
+        missing_layers = set()
+
+        for potential_layer in graph.find_children(container):
+            _, _, layer_name = potential_layer.rpartition(".")
+            if layer_name not in expected_layers:
+                missing_layers.add(layer_name)
+
+        if missing_layers:
+            # Sorting isn't strictly neccessary, but it makes things predictable for testing
+            layers = ", ".join(sorted(f"'{name}'" for name in missing_layers))
+            # This message is gramatically bad when there is a single missing layer but it should
+            # still make sense
+            raise ValueError(
+                f"Missing potential layers from container '{container}': "
```

Comment:
I think the idea of a 'potential' layer is a bit confusing. How about a message like:

"Unexpected module found in an exhaustive contract. To disable this check, either set exhaustive = false, or add the module to the exhaustive_ignores. Unexpected module: mypackage.foo."

Plural version:

"Unexpected module found in an exhaustive contract. To disable this check, either set exhaustive = false, or add the modules to the exhaustive_ignores. Unexpected modules: mypackage.foo, mypackage.bar."

Reply:
Incidentally, another approach we could take here is to treat it not as an error, but as a contract failure. In that case, instead of raising an exception (which aborts everything) we could add the unexpected modules to the ContractCheck metadata that is returned by the check method. That would then allow a richer reporting on the unexpected modules, in combination with other failures (in this case, invalid chains).

The other advantage of that approach is that if there were multiple contracts in the check, we would not hard fail on all of them - instead we'd just see this issue alongside any other passing/failing contracts.

I don't mind too much doing it this way, but I thought worth raising as an alternative.